ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+      (a : (a, allowed * r) mode) (b : (a, l * allowed) mode) =
+    let log = if logging then Some (ref []) else None in
+    let submode_cc left right =
+      if C.le obj left right then Ok () else Error { left; right }
+    in
+    let submode_mvc v right =
+      Result.map_error
+        (fun left ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_mvc ?log obj v right)
+    in
+    let submode_cmv left v =
+      Result.map_error
+        (fun right ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_cmv ?log obj left v)
+    in
+    let submode_mvmv v u =
+      Result.map_error
+        (fun (left, right) ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_mvmv ?log obj v u)
+    in
+    let r =
+      match a, b with
+      | Amode left, Amode right -> submode_cc left right
+      | Amodevar v, Amode right -> submode_mvc v right
+      | Amode left, Amodevar v -> submode_cmv left v
+      | Amodevar v, Amodevar u -> submode_mvmv v u
+      | Amode a, Amodemeet (b, mvs) ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv -> submode_cmv a mv) mvs)
+      | Amodevar mv, Amodemeet (b, mvs) ->
+        Result.bind (submode_mvc mv b) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv mv') mvs)
+      | Amodejoin (a, mvs), Amode b ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv' -> submode_mvc mv' b) mvs)
+      | Amodejoin (a, mvs), Amodevar mv ->
+        Result.bind (submode_cmv a mv) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv' mv) mvs)
+      | Amodejoin (a, mvs), Amodemeet (b, mus) ->
+        (* TODO: mabye create a intermediate variable? *)
+        Result.bind (submode_cc a b) (fun () ->
+            Result.bind
+              (find_error (fun mv -> submode_mvc mv b) mvs)
+              (fun () ->
+                Result.bind
+                  (find_error (fun mu -> submode_cmv a mu) mus)
+                  (fun () ->
+                    find_error
+                      (fun mu -> find_error (fun mv -> submode_mvmv mv mu) mvs)
+                      mus)))
+    in
+    match r with
+    | Ok () -> Ok log
+    | Error e ->
+      (* we mutated some states and found conflict;
+         need to revert those mutation to keep the state consistent.
+         A nice by-effect is that this function doesn't mutate state in failure
+      *)
+      Option.iter (fun log -> undo_changes !log) log;
+      Error e
+
+  let submode obj a b =
+    match submode_try obj a b with
+    | Ok log ->
+      Option.iter !append_changes log;
+      Ok ()
+    | Error _ as e -> e
+
+  let zap_to_ceil_morphvar obj mv =
+    assert (submode obj (Amode (mupper obj mv)) (Amodevar mv) |> Result.is_ok);
+    mupper obj mv
+
+  let zap_to_ceil : type a l. a C.obj -> (a, l * allowed) mode -> a =
```

Comment:
I like this implementation. Can `zap_to_floor` be done similarly? If not, why not?

Reply:
The `upper` field is precise, so to zap `v` to ceil, we simply call `submode v.upper v`.

The `lower` field is conservative, so calling `submode v v.lower` might error - because the real lower bound of `v` might be higher and not reflected by `v.lower`.